salt: Add a yum package dependency manager

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt', 'downgrade'

**Context**: 

Issue: #1143

**Summary**:

We would like to support package downgrade plus their corresponding dependencies. 
So what have we done?
- Introducing a way to find packages + dependencies (supported on redhat only for now) using salt module.
- Adapting pkg.installed to work with our custom module.

**Acceptance criteria**: 

- Be able to downgrade a package like Kubelet successfully
- Used in  parallel with Downgrade orchestrate, we should be able downgrade metalk8s to a previous minor version.
- used to test : #1343


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1143

<!-- If you want to refer to an issue while not closing it, use:



-->

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1143-add-downgrade-package-manager` into
`development/2.0`
* [w/2.1/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1143-add-downgrade-package-manager) will be merged into `development/2.1` 
* [w/2.2/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1143-add-downgrade-package-manager) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

NicolasT commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,55 @@
+    try:
+        os_family = __grains__['os_family'].lower()
+    except Exception:
+        return (False, "No yum based system detected")
```
The above checks whether there's an `os_family` grain, which says nothing about being YUM based or not?

Ebaneck replied:
my bad!

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,55 @@
+    try:
+        os_family = __grains__['os_family'].lower()
+    except Exception:
+        return (False, "No yum based system detected")
```
Hm, no the `try` definitly not check that

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,55 @@
+
+    if os_family == 'redhat':
+        return __virtualname__
+    return(False, "Metalk8s_package_manager: no yum based system detected")
```
```suggestion
    return (False, "metalk8s_package_manager: no yum based system detected")
```

TeddyAndrieux replied:
If you prefix here you should do it every where

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,55 @@
+        return (False, "No yum based system detected")
+
+    if 'metalk8s_package_manager.list_pkg_deps' not in __salt__:
+        return (False, '`metalk8s_package_manager.list_pkgs_deps` not found')
```
```suggestion
        return (False, "metalk8s_package_manager: `list_pkgs_deps` not found")
```

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,55 @@
+    return(False, "Metalk8s_package_manager: no yum based system detected")
+
+
+def pkg_deps(name, version):
```
This is definitly not a state. 
The state should get the pkg_deps using the module and use it directly in a classic `pkg.installed`

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,12 +2,16 @@
 {%- macro pkg_installed(name='') -%}
   {%- set package = repo.packages[name] | default({}) %}
+  {%- set version = package.version | default(None) %}
+  {%- set deps_list = salt.metalk8s_package_manager.list_pkg_deps(name, version) %}
```
You can't do it in jinja.
As when the jinja is rendered, the repository may not have been configured yet.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# History mismatch



Merge commit #ccc3520c61bc3c7442f9d11f075db91bc7d5bd9e on the integration branch
`w/2.1/feature/1143-add-downgrade-package-manager` is merging a branch which is neither the current
branch `feature/1143-add-downgrade-package-manager` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1143-add-downgrade-package-manager` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,54 @@
+
+    if __grains__['os_family'].lower() == 'redhat':
+        return __virtualname__
+    return (False, "Metalk8s_package_manager: no yum based system detected")
```
Minor: here you should use `__virtualname__` in the error message IMO

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,12 +2,16 @@
     - update_holds: True
-    {%- endif %}
+    {% endif %}
+    {%- endfor %}
```
```suggestion
  {%- endfor %}
```

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,12 +2,16 @@
     - hold: True
     - update_holds: True
-    {%- endif %}
+    {% endif %}
```
```suggestion
    {%- endif %}
```

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,12 +2,16 @@
     - fromrepo: {{ repo.repositories.keys() | join(',') }}
-    {%- if package.version | default(None) %}
-    - version: {{ package.version }}
+    {% if version  %}
```
```suggestion
    {%- if version  %}
```

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,12 +2,16 @@
   {%- set package = repo.packages[name] | default({}) %}
+  {%- set version = package.version | default(None) %}
+  {%- set deps_list = salt.metalk8s_package_manager.list_pkg_deps(name, version) %}
+  {%- for name, version in deps_list.items() %}
```
Also, doing the `for` here will never work. Because you will create one `pkg.installed` per deps it's not what we want. We want to give all deps in one command so that salt can use them to one `yum` command

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,56 @@
+    deps_list = __salt__['cmd.run_all'](
+        [
+            'repoquery', '-q', '--qf', '%{NAME} %{VERSION}', '--installed',
+            '--whatrequires', '{}-{}'.format(str(pkg_name), str(version))
```
Hmm for me. That doesn't works as you need to enable the repo `--enablerepo=...` and also the package is holded here so you will always get empty answer.
Also `--installed` doesn't work on my side. Always got empty answer (btw it's works since we use kubelet-1.11.10 that require the same version of kubernetes-cni as the kubelet 1.12.6

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,79 @@
+so that we can support downgrade in metalk8s
+'''
+
+import re
```
No longer needed

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,79 @@
+import re
+import logging
+
+from salt.exceptions import CommandExecutionError
```
Not used

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,39 @@
+
+import logging
+
+from salt.exceptions import CommandExecutionError
```
Ditto

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,79 @@
+    if not version:
+        return {pkg_name: None}
+
+    pkgs_list = {pkg_name: version}
```
I think you can refactor this 3 line in 
```
pkgs_list = {pkg_name: version}

if not version:
  return pkgs_list
```

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,79 @@
+        command_all.append('--enablerepo="{}"'.format(fromrepo))
+
+    deps_list = __salt__['cmd.run_all'](command_all)
+    log.debug(deps_list)
```
I think you can remove this one

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,79 @@
+                log.error(
+                    'Failed to check if package is installed: %s',
+                    deps_list['stderr'] or deps_list['stdout']
+                )
```
You need to `return None` also here

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,39 @@
+
+
+def __virtualname__():
+
```
One empty line

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,39 @@
+    out = []
+    for pkg, version in dep_list.items():
+        pkg_data = {pkg: version}
+        out.append(pkg_data)
```
This for will basically just copy the dep_list ? 
This is useless IMO just give `dep_list` directly to `pkg.installed`

jbertran replied:
You're building a `{pkg: version}` list of dicts with only one key/value pair from a dict with a bunch of keys and values.

TeddyAndrieux replied:
Oh yes it's true MB

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,12 +2,12 @@
     - name: {{ name }}
     - fromrepo: {{ repo.repositories.keys() | join(',') }}
-    {%- if package.version | default(None) %}
+    {% if package.version %}
```
Don't remove the `default(None)` as maybe `packages` object can be empty

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,12 +2,12 @@
     - hold: True
     - update_holds: True
-    {%- endif %}
+    {% endif %}
```
Do not remove the `-`

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1143-add-downgrade-package-manager` into
`development/2.0`
* [w/2.1/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1143-add-downgrade-package-manager) will be merged into `development/2.1` 
* [w/2.2/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1143-add-downgrade-package-manager) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,76 @@
+    return __virtualname__
+
+
+def list_pkg_deps(pkg_name, version, fromrepo=None):
```
Version should default to `None`

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,76 @@
+            'Failed to list package dependencies: %s',
+            deps_list['stderr'] or deps_list['stdout']
+        )
+        return None
```
Minor: add one empty line just after the return :)

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,76 @@
+                    'Failed to check if package is installed: %s',
+                    deps_list['stderr'] or deps_list['stdout']
+                )
+                return None
```
Minor: One line also here please

TeddyAndrieux reviewed (commented):
Some minor comment and then should be ok

bert-e commented:
# History mismatch



Merge commit #16c8f6e319ab42c5becdec8be9e3d1a756e03ae4 on the integration branch
`w/2.1/feature/1143-add-downgrade-package-manager` is merging a branch which is neither the current
branch `feature/1143-add-downgrade-package-manager` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1143-add-downgrade-package-manager` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1143-add-downgrade-package-manager` into
`development/2.0`
* [w/2.1/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1143-add-downgrade-package-manager) will be merged into `development/2.1` 
* [w/2.2/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1143-add-downgrade-package-manager) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

bert-e commented:
# History mismatch



Merge commit #faad7c76caa978e745dd05b3ac1465e66c04ad5b on the integration branch
`w/2.1/feature/1143-add-downgrade-package-manager` is merging a branch which is neither the current
branch `feature/1143-add-downgrade-package-manager` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1143-add-downgrade-package-manager` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,36 @@
+    return (False, "metalk8s_package_manager: no yum based system detected")
+
+
+def install_with_deps(name, version=None, fromrepo=None, **kwargs):
```
Maybe you should rename this one `installed`

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,36 @@
+        pkg_data = {pkg: version}
+        out.append(pkg_data)
+    return __states__["pkg.installed"](
+        name=name, pkgs=out, fromrepo=fromrepo, **kwargs
```
If you want a one liner instead of the for you can do
```suggestion
        name=name, pkgs=[{p: v} for p, v in dep_list.items()], fromrepo=fromrepo, **kwargs
```

slaperche-scality replied:
The list comprehension is a good suggestion, but I would keep it out of the function call to improve the readability:

```python
    pkgs = [{name: version} for name, version in dep_list.items()]
    return __states__["pkg.installed"](
        name=name, pkgs=pkgs, fromrepo=fromrepo, **kwargs
    )
```

TeddyAndrieux reviewed (changes requested):
Seems you have a build issue also

TeddyAndrieux commented on scripts/bootstrap.sh.in:
```diff
@@ -299,6 +299,8 @@ orchestrate_bootstrap() {
             pillar="${pillar[*]}"
 
+    sleep 10
+
```
You can remove this one now. Was only for testing purpose

TeddyAndrieux commented on salt/metalk8s/macro.sls:
```diff
@@ -2,10 +2,10 @@
     - name: {{ name }}
     - fromrepo: {{ repo.repositories.keys() | join(',') }}
-    {%- if package.version | default(None) %}
+    {% if package.version | default(None) %}
```
Do not remove the `-` :)

gdemonet replied:
FYI: http://jinja.pocoo.org/docs/2.10/templates/#whitespace-control

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,78 @@
+    for line in out:
+        name, version = line.strip().split()
+        pkgs_list[name] = version
+        for key, value in pkgs_list.items():
```
value not used
```suggestion
        for key, _ in pkgs_list.items():
```

gdemonet replied:
That loop should not be performed for each new dependency: just do it once after the `for line in out:` loop.

gdemonet commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,78 @@
+
+    out = deps_list['stdout'].splitlines()
+    for line in out:
+        name, version = line.strip().split()
```
Would be nice to have some error management, with some `try:`/`except ValueError:` in case some lines are not in the expected format. (minor, since you can rely on `repoquery` to have a constant output format).

gdemonet commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,78 @@
+        str(pkg_name),
+        str(version)
+    )
+    pkgs_list = {pkg_name: version}
```
So, `pkgs_list` is actually a dictionary...? Could do with a more explicit name IMHO.

gdemonet commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,78 @@
+        return pkgs_list
+
+    command_all = [
+        'repoquery', '--whatrequires', '--qf',
```
Need to add `--recursive`, just in case :)

gdemonet commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,78 @@
+    out = deps_list['stdout'].splitlines()
+    for line in out:
+        name, version = line.strip().split()
+        pkgs_list[name] = version
```
To anticipate conflicts, it could be interesting to check if `name` is already in `pkgs_list` and compare versions in such situation

gdemonet commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,33 @@
+def __virtualname__():
+    if __grains__['os_family'].lower() == 'redhat':
+        return __virtualname__
+    return (False, "metalk8s_package_manager: no yum based system detected")
```
Let's not create artificial reasons: just state that this state only works on RedHat family of OSes.

slaperche-scality replied:
Or simply `RPM-based system`

gdemonet reviewed (dismissed):
Good work, let's improve the `list_pkg_deps` module and we should be good to go :)

Ebaneck committed:
salt: Add a package dependency manager

This commit adds a salt module that manages
yum packages and their dependencies.

Full dependency list is required to support yum package
downgrade properly. We have also replaced the `pkg.installed`
that rely on macro.sls to use the custom state.

TeddyAndrieux commented on salt/_modules/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,79 @@
+
+    for key in pkgs_dict.keys():
+        package_query = __salt__['cmd.run_all'](
+            ['rpm', '-qa', key], python_shell=False
```
Why disable `python_shell` ?

TeddyAndrieux commented on salt/_states/metalk8s_package_manager.py:
```diff
@@ -0,0 +1,33 @@
+    dep_list = __salt__['metalk8s_package_manager.list_pkg_deps'](
+        name, version, fromrepo
+    )
+    pkgs = [{name: version} for k, v in dep_list.items()]
```
It's wrong you need to use `k` and `v` here btw
```suggestion
    pkgs = [{k: v} for k, v in dep_list.items()]
```

slaperche-scality replied:
Better fix is to keep meaningful name IMHO…

```python
pkgs = [{name: version} for name, version in dep_list.items()]
```

TeddyAndrieux commented on salt/metalk8s/orchestrate/bootstrap/init.sls:
```diff
@@ -74,6 +74,8 @@ Sync bootstrap minion:
   - name: saltutil.sync_all
   - tgt: {{ pillar.bootstrap_id }}
   - saltenv: metalk8s-{{ version }}
```
Remove this line

TeddyAndrieux commented on salt/metalk8s/orchestrate/bootstrap/init.sls:
```diff
@@ -73,7 +73,8 @@ Sync bootstrap minion:
   - tgt: {{ pillar.bootstrap_id }}
-  - saltenv: metalk8s-{{ version }}
+  - kwarg:
+      saltenv: metalK8s-{{ version }}
```
Remove capital `K` here. No capital

Ebaneck replied:
ok

Ebaneck commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1143-add-downgrade-package-manager` into
`development/2.0`
* [w/2.1/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1143-add-downgrade-package-manager) will be merged into `development/2.1` 
* [w/2.2/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1143-add-downgrade-package-manager) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# History mismatch



Merge commit #ea1c1baee0151fe074cc0833be74260575a0072e on the integration branch
`w/2.1/feature/1143-add-downgrade-package-manager` is merging a branch which is neither the current
branch `feature/1143-add-downgrade-package-manager` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/1143-add-downgrade-package-manager` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/1143-add-downgrade-package-manager` into
`development/2.0`
* [w/2.1/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/1143-add-downgrade-package-manager) will be merged into `development/2.1` 
* [w/2.2/feature/1143-add-downgrade-package-manager](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/1143-add-downgrade-package-manager) will be merged into `development/2.2` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @nootal

* @cmonfort

* @lucieleonard

* @alxf

* @carlito-scality

* @alexandre-allard-scality

* @teddyandrieux

* @slaperche-scality

* @thomasdanan

* @zempashi

* @nicolast

* @sayf-eddine-scality

* @monpote

* @gdemonet

* @tcarmet

Ebaneck commented:
@bert-e approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye ebaneck.
